File: test/Java-perf-application/src/main/java/zebrogamq/perf/j2se/GameLogicApplication.java
/**
 ZebroGaMQ: Communication Middleware for Mobile Gaming
 Copyright: Copyright (C) 2009-2012
 Contact: denis.conan@telecom-sudparis.eu, michel.simatic@telecom-sudparis.eu

 This library is free software; you can redistribute it and/or
 modify it under the terms of the GNU Lesser General Public
 License as published by the Free Software Foundation; either
 version 3 of the License, or any later version.

 This library is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 Lesser General Public License for more details.

 You should have received a copy of the GNU Lesser General Public
 License along with this library; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307
 USA

 Developer(s): Denis Conan, Gabriel Adgeg
 */

package zebrogamq.perf.j2se;

import java.io.IOException;
import java.io.InputStream;
import java.util.Properties;
import java.util.Random;
import java.lang.StringBuffer;

import zebrogamq.gamelogic.ChannelsManager;
import zebrogamq.gamelogic.GameLogicState;
import zebrogamq.gamelogic.JoinAction;
import zebrogamq.gamelogic.OptionalDelegationOfStandardActions;
import zebrogamq.gamelogic.PresenceAction;
import zebrogamq.gamelogic.Util;

public class GameLogicApplication {

	private static MyGameLogicState state = null;
	private static MyOptionalDelegationOfStandardActions modosa = null;
	private static Random rand = null;

	public static void main(final String[] argv) {
		//Util.setLogger(new Logger());
		if (argv.length != 8) {
			Util.println("Please, only eight strings"
					+ " are accepted "
					+ "<login> <password> <role> <game name> <instance name> <nb msgs> <millisec between msg> <payload size>.");
			return;
		}
		boolean loadOK = loadProperties();
		if (loadOK) {
			// instantiate GameLogicState
			state = new MyGameLogicState();
			state.login = argv[0];
			state.password = argv[1];
			state.role = argv[2];
			state.gameName = argv[3];
			state.instanceName = argv[4];
			state.nbMsgs = Integer.parseInt(argv[5]);
			state.waitBetweenMsgs = Integer.parseInt(argv[6]);
			state.payloadSize = Integer.parseInt(argv[7]);
			
			rand = new Random(System.currentTimeMillis());
			// execute the XMLRPC call
			boolean loggedIn = executeXMLRPCLogin();
			if (loggedIn) {
				initChannelsManager();
				// launch the participant list thread (except for spectators
				// applications
				if (!state.role.equals(GameLogicState.SPECTATOR)) {
					startParticipantMsgThread();
				}
			} else {
				Util.println("Bad XML-RPC answer.");
			}
		} else {
			Util.println("Properties files have not been load.");
		}
	}

	private static boolean loadProperties() {
		boolean loadOK = false;
		Properties rabbitMQProperties = new Properties();
		String rabbitMQConfigFileName = System
				.getProperty("rabbitmq.config.file");
		InputStream in = Util.class.getClassLoader().getResourceAsStream(
				rabbitMQConfigFileName);
		if (in == null) {
			Util.println("RabbitMQ configuration file \""
					+ rabbitMQConfigFileName + "\" not found");
			throw new RuntimeException();
		}
		try {
			rabbitMQProperties.load(in);
			Util.setRabbitMQProperties(rabbitMQProperties);
		} catch (IOException e) {
			e.printStackTrace();
		}
		Properties xmlrpcProperties = new Properties();
		String xmlrpcConfigFileName = System.getProperty("xmlrpc.config.file");
		in = Util.class.getClassLoader().getResourceAsStream(
				xmlrpcConfigFileName);
		if (in == null) {
			Util.println("XMLRPC configuration file \"" + xmlrpcConfigFileName
					+ "\" not found");
			throw new RuntimeException();
		}
		try {
			xmlrpcProperties.load(in);
			Util.setXMLRPCProperties(xmlrpcProperties);
			loadOK = true;
		} catch (IOException e) {
			e.printStackTrace();
		}
		return loadOK;
	}

	private static boolean executeXMLRPCLogin() {
		boolean res = false;
		if (state.role.equals(GameLogicState.GAME_MASTER)) {
			Util.println("[" + state.role + " " + state.login
					+ "] creating instance " + state.gameName + "/"
					+ state.instanceName);
			res = XMLRPCLogin.createAndJoinGameInstance(state.login,
					state.password, state.gameName, state.instanceName);
		} else if (state.observationKey != null) {
			Util.println("[" + state.role + " " + state.login
					+ "] joining instance " + state.gameName + "/"
					+ state.instanceName + " with observation key "
					+ state.observationKey);
			res = XMLRPCLogin.joinGameInstance(state.login, state.password,
					state.gameName, state.instanceName, state.observationKey);
		} else {
			Util.println("[" + state.role + " " + state.login
					+ "] joining instance " + state.gameName + "/"
					+ state.instanceName);
			res = XMLRPCLogin.joinGameInstance(state.login, state.password,
					state.gameName, state.instanceName);
		}
		return res;
	}

	private static boolean executeXMLRPCTerminateGameInstance() {
		boolean res = false;
		return res;
	}

	/*
	 * This method successively:
	 * - instantiate the channelsManager of the state in order to 
	 *   start the consumption and to enable publication of messages.
	 * - publish a JOIN message to the GameLogicServer. By default,
	 *   on the reception of this message, there is only a message
	 *   displayed. If you want to define your own behavior on the 
	 *   reception of this message, please refer to the javadoc of 
	 *   the class OptionalDelegationOfStandardActions.
	 *   (http://simatic.github.com/ZebroGaMQ/doc/javadoc/zebrogamq/gamelogic/OptionalDelegationOfStandardActions.html)
	 */
	private static void initChannelsManager() {
		try {
			// Instantiate the channelsManager
			state.channelsManager = ChannelsManager.getInstance(state,
					MyListOfGameLogicActions.ListOfActionsMaps);
			// Publish a Join message
			String content = state.login + "," + state.gameName + ","
					+ state.instanceName;
			state.channelsManager.publishToGameLogicServer(state,
					JoinAction.JOIN, content);
		} catch (IOException e) {
			e.printStackTrace();
		}
		modosa = new MyOptionalDelegationOfStandardActions();
		OptionalDelegationOfStandardActions.setInstance(modosa);
	}

	private static String randomParticipant(){
		// Participants is structured as follows: {'participant_1': 'Thu Jul 12 10:45:59 CEST 2012', 'participant_2': 'Thu Jul 12 10:45:57 CEST 2012'}
		String participantsWithoutBrackets = modosa.participants.substring(1, modosa.participants.length()-1);
		String [] participantsArray = participantsWithoutBrackets.split(",");
		String participantWithDate = participantsArray[rand.nextInt(participantsArray.length)].trim();
		String [] part = participantWithDate.split("'");
		return part[1];
	}
	
	private static void startParticipantMsgThread() {
		new Thread() {
			public void run() {
				int sentMsgs = 0;
				StringBuffer sb = new StringBuffer(state.payloadSize);
				for (int i = 0; i < state.payloadSize; i++) {
					sb.append('A');					
				}
				long time = System.currentTimeMillis();
				String s;
				String consumer;
				while (!state.hasConnectionExited() && (sentMsgs < state.nbMsgs)) {
					time = System.currentTimeMillis();
					s = time + ",";
					try {
						switch (sentMsgs%4) {
						case 0:
							state.channelsManager.publishToGameLogicServer(state,
									PresenceAction.ASK_PARTICIPANTS_LIST, "");
							Stat.addSentPingServer(0);
							break;
						case 1:
							consumer = randomParticipant(); 
							s += consumer + ",";
							if (s.length() < state.payloadSize)
								s += sb.substring(s.length());
							state.channelsManager.publishToAll(state,
									PingActionKind.ALL_ACTION, s);
							Stat.addSentPingAll(s.length());
							break;
						case 2:
							consumer = randomParticipant(); 
							if (s.length() < state.payloadSize)
								s += sb.substring(s.length());
							state.channelsManager.publish(consumer, state,
									PingActionKind.PLAYER_ACTION, s);							
							Stat.addSentPingPlayer(s.length());
							break;
						default:
							if (s.length() < state.payloadSize)
								s += sb.substring(s.length());
							state.channelsManager.publishToGameLogicServer(state,
									PingActionKind.SERVER_ACTION, s);
							Stat.addSentPingServer(s.length());
							break;
						}
					} catch (IOException e) {
						e.printStackTrace();
					}
					sentMsgs += 1;
					// wait for a while
					try {
						Thread.sleep(state.waitBetweenMsgs);
					} catch (InterruptedException e) {
						Util.println(" [Player " + state.login
								+ "] Thread sleep was interrupted");
					}
				}
				// If we are the GAME_MASTER, we terminate the Game instance
				if (state.role.equals(GameLogicState.GAME_MASTER)) {
					Util.println("[" + state.role + " " + state.login
							+ "] terminating instance " + state.gameName + "/"
							+ state.instanceName);
					if (!XMLRPCLogin.terminateGameInstance(state.gameName, state.instanceName)){
						Util.println("***Problem with XMLRPCLogin.terminateGameInstance: Must exit!***");
						Stat.results();
						System.exit(1);				
					}
				}
				// We could do a System.exit(0) here, but it is more interesting to do it in terminate method of 
				// MyOptionalDelegationOfStandardActions. By doing so, all participants stop when the Game_MASTER
				// terminates
				//System.exit(0);				
			}
		}.start();
	}
}

File: test/ios-application/ios-application/src/GameLogicApplication.cpp
/**
 ZebroGaMQ: Communication Middleware for Mobile Gaming
 Copyright: Copyright (C) 2009-2012
 Contact: denis.conan@telecom-sudparis.eu, michel.simatic@telecom-sudparis.eu

 This library is free software; you can redistribute it and/or
 modify it under the terms of the GNU Lesser General Public
 License as published by the Free Software Foundation; either
 version 3 of the License, or any later version.

 This library is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 Lesser General Public License for more details.

 You should have received a copy of the GNU Lesser General Public
 License along with this library; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307
 USA

 Developer(s): Michel SIMATIC, Van Hung LE
 */

#include "GameLogicApplication.h"
#include "XMLRPCLogin.h"
#include <ctime>
#include <iostream>
#include <thread>
#include <chrono>
#include <stdio.h>
#include <stdlib.h>
#include "MyListOfGameLogicActions.h"
#include <zebrogamq-ios/ChannelsManager.h>
#include <zebrogamq-ios/JoinAction.h>
#include <zebrogamq-ios/PresenceAction.h>
#include <zebrogamq-ios/ZebroGamqUtil.h>
#include <fstream>

#include <unistd.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netdb.h>
#include <errno.h>
#include <fcntl.h>

using namespace std;

GameLogicState* GameLogicApplication::state = 0;
std::string GameLogicApplication::XMLRPC_PROPERTIES_FILE = "./resources/xmlrpc.properties";
std::string GameLogicApplication::CONFIG_PROPERTIES_FILE = "./resources/config.properties";
std::string GameLogicApplication::RABBITMQ_PROPERTIES_FILE = "./resources/rabbitmq.properties";

bool GameLogicApplication::executeXMLRPCLogin(std::string action) {
    if (action.compare("createAndJoinGameInstance")==0) {
        return XMLRPCLogin::createAndJoinGameInstance(GameLogicApplication::state->login, GameLogicApplication::state->password, GameLogicApplication::state->gameName, GameLogicApplication::state->instanceName);
    }
    if (action.compare("joinGameInstance")==0) {
        return XMLRPCLogin::joinGameInstance(GameLogicApplication::state->login, GameLogicApplication::state->password, GameLogicApplication::state->gameName, GameLogicApplication::state->instanceName);
    }
    return nullptr;
}

/*
 * This method successively:
 * - instantiate the channelsManager of the GameLogicApplication::state in order to
 *   start the consumption and to enable publication of messages.
 * - publish a JOIN message to the GameLogicServer. By default,
 *   on the reception of this message, there is only a message
 *   displayed. If you want to define your own behavior on the
 *   reception of this message, please refer to the javadoc of
 *   the class OptionalDelegationOfStandardActions.
 *   (http://simatic.github.com/ZebroGaMQ/doc/javadoc/zebrogamq/gamelogic/OptionalDelegationOfStandardActions.html)
 */
void GameLogicApplication::initChannelsManager() {
	MyListOfGameLogicActions* myListOfGameLogicActions = new MyListOfGameLogicActions();
	GameLogicApplication::state->channelsManager = ChannelsManager::getInstance(GameLogicApplication::state, myListOfGameLogicActions->m_list);
	// Publish a Join message
	std::string content = GameLogicApplication::state->login + "," + GameLogicApplication::state->gameName + "," + GameLogicApplication::state->instanceName;
	state->channelsManager->publishToGameLogicServer(state, JOIN, content);
}

void GameLogicApplication::startParticipantListThread() {
	std::thread thread (threadStartParticipantListThread);
	thread.detach();
}

void GameLogicApplication::threadStartParticipantListThread() {
	while (!GameLogicApplication::state->hasConnectionExited()) {
        state->channelsManager->publishToGameLogicServer(state, ASK_PARTICIPANTS_LIST, " ");
		// wait for a while
		usleep(15000 * 1000);
	}
}

/*
* TODO: Load properties files
*/
bool GameLogicApplication::loadProperties() {
	bool result = false;
	Properties* xmlrpcProperty = new Properties();
	Properties* configProperty = new Properties();
	Properties* rabbitmqProperty = new Properties();

	result = xmlrpcProperty->load(GameLogicApplication::XMLRPC_PROPERTIES_FILE);
	if (!result) return false;

	result = configProperty->load(GameLogicApplication::CONFIG_PROPERTIES_FILE);
	if (!result) return false;

	result = rabbitmqProperty->load(GameLogicApplication::RABBITMQ_PROPERTIES_FILE);
	if (!result) return false;

	ZebroGamqUtil::setRabbitMQProperties(rabbitmqProperty);
	ZebroGamqUtil::setConfigProperties(configProperty);
	ZebroGamqUtil::setXMLRPCProperties(xmlrpcProperty);

	return result;
}

/*
 * TODO: Update config properties
 */
bool GameLogicApplication::updateConfigProperties(std::string propType, std::string propValue, const std::string &propFilename) {
    bool result = true;
    FILE *f = fopen(propFilename.c_str(), "w");
    if (f == nullptr) return false;
    fprintf(f, "%s    %s", propType.c_str(), propValue.c_str());
    fclose(f);
    return result;
}

/*
 * TODO: Load properties files
 */
bool GameLogicApplication::loadProperties(std::string xmlrpcPropertiesFile, std::string rabbitmqPropertiesFile, std::string configPropertiesFile) {
	bool result = false;
	Properties* xmlrpcProperty = new Properties();
	Properties* rabbitmqProperty = new Properties();
    Properties* configProperty = new Properties();
    
	result = xmlrpcProperty->load(xmlrpcPropertiesFile);
	if (!result) return false;
    
	result = rabbitmqProperty->load(rabbitmqPropertiesFile);
	if (!result) return false;
    
    result = configProperty->load(configPropertiesFile);
	if (!result) return false;
    
	ZebroGamqUtil::setRabbitMQProperties(rabbitmqProperty);
	ZebroGamqUtil::setXMLRPCProperties(xmlrpcProperty);
    ZebroGamqUtil::setConfigProperties(configProperty);
    
	return result;
}

/* Method check the XMLRPC server is available in (host, port) or not */
bool GameLogicApplication::checkXMLRPCServer() {
    std::string GAME_SERVER_URL = ZebroGamqUtil::getXMLRPCProperties()->getProperty("gameServerXMLRPCHost");
	int GAME_SERVER_PORT = atoi(ZebroGamqUtil::getXMLRPCProperties()->getProperty("gameServerXMLRPCPort").c_str());
	int MAX_RETRY = atoi(ZebroGamqUtil::getXMLRPCProperties()->getProperty("maxRetry").c_str());
    
    struct sockaddr_in saddr;
    memset(&saddr, 0, sizeof(saddr));
    saddr.sin_family = AF_INET;
    
    struct hostent *hp = gethostbyname(GAME_SERVER_URL.c_str());
    if (hp == 0) return false;
    
    saddr.sin_family = hp->h_addrtype;
    memcpy(&saddr.sin_addr, hp->h_addr, hp->h_length);
    saddr.sin_port = htons((u_short) GAME_SERVER_PORT);

    int result = -1;
    for(int i = 0; i < MAX_RETRY; i++) {
    	// For asynch operation, this will return EWOULDBLOCK (windows) or
    	// EINPROGRESS (linux) and we just need to wait for the socket to be writable...
        int sock = ::socket(AF_INET, SOCK_STREAM, 0);
        result = ::connect(sock, (struct sockaddr *)&saddr, sizeof(saddr));
        if (result == 0) {
            close(sock);
            break;
        }
        close(sock);
        usleep(1000 * 1000);
    }
    return (result == 0);
}